keps/sig-node/20200619-provisioning-resources-with-numa-topology.md

```diff
@@ -0,0 +1,240 @@
+
+
+### Design based on podresources interface of the kubelet
+
+Podresources interface of kubelet described in
+
+pkg/kubelet/apis/podresources/v1alpha1/api.proto
+
+it available for every process on the worker node by
+unix domain socket situated by the following path:
+
+filepath.Join(kl.getRootDir(), config.DefaultKubeletPodResourcesDirName)
+
+it could be used to collect used resources on the worker node and to evaluate
+its NUMA assignment (by device id)
+
+For example
+TODO add path on the real system
+
+Currently it doesn't provide information from CPUManager only from
+DeviceManager as well as information about memory which was used.
+But information provided by podresource could be easily extended
+without modification of API.
+
+### API
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document](https://docs.google.com/document/d/12kj3fK8boNuPNqob6F_pPU9ZTaNEnPGaXEooW1Cilwg/edit).
+
+
+```go
+// NodeResourceTopology is a specification for a Foo resource
+type NodeResourceTopology struct {
+       metav1.TypeMeta   `json:",inline"`
+       metav1.ObjectMeta `json:"metadata,omitempty"`
+
+       TopologyPolicy string
+       Nodes   []NUMANodeResource   `json:"nodes"`
+}
+
+// NUMANodeResource is the spec for a NodeResourceTopology resource
+type NUMANodeResource struct {
+       NUMAID int
+       Resources v1.ResourceList
+}
+```
+
+The code for working with it is generated by https://github.com/kubernetes/code-generator.git
+One CRD instance contains information of available resources of the appropriate worker node.
+
+In order to allow the Daemon (running as a pod on each node in the cluster) to create, get, update, delete NodeResourceTopology CRD instances, ClusterRole and ClusterRoleBinding would have to be configured as below:
+
+``` yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: noderesourcetopology-handler
+rules:
+- apiGroups: ["k8s.cncf.io"]
+  resources: ["noderesourcetopologies"]
+  verbs: ["*"]
+- apiGroups: ["rbac.authorization.k8s.io"]
+  resources: ["*"]
+  verbs: ["*"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: handle-noderesourcetopology
+subjects:
+- kind: ServiceAccount
+  name: noderesourcetopology-account
+  namespace: default
+roleRef:
+  kind: ClusterRole
+  name: noderesourcetopology-handler
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: noderesourcetopology-account
+```
+
+`serviceAccountName: noderesourcetopology-account` would have to be added to the manifest file of the Daemon.
+
+### Graduation Criteria
+
+* The feature has been stable and reliable in the past several releases.
+* Documentation should exist for the feature.
+* Test coverage of the feature is acceptable.
```

Comment:
what is the recommended test strategy?
what resources are needed in the project to sustainable maintain the signals they produce.

Reply:
The same requirements for hardware as for TopologyManager, but maybe here is better to have at least 2 nodes with baremetal installation or VM with NUMA topology. I'll describe it in details here.